File: solutions/scala/153-insert-delete-getrandom-o-1.scala
/**
 * Problem 153: Insert Delete GetRandom O(1) (LeetCode 380)
 * Difficulty: Med
 * Language: Scala
 */
import scala.util.Random
class RandomizedSet {
  private var map = Map.empty[Int, Int]
  private var list = Vector.empty[Int]
  def insert(v: Int): Boolean =
    if (map.contains(v)) false
    else { map += (v -> list.size); list :+= v; true }
  def remove(v: Int): Boolean =
    if (!map.contains(v)) false
    else {
      val idx = map(v); val last = list.last
      val nl = list.init.updated(idx, last)
      map = (map - v).updated(last, idx)
      if (idx == list.size - 1) map -= v
      list = if (idx < nl.size) nl else list.init
      true
    }
  def getRandom(): Int = list(Random.nextInt(list.size))
}

File: solutions/clojure/153-insert-delete-getrandom-o-1.clj
;; Problem 153: Insert Delete GetRandom O(1) (LeetCode 380)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn create-set [] {:map {} :list []})
(defn rs-insert [{:keys [map list]} val]
  (if (contains? map val) [false {:map map :list list}]
    [true {:map (assoc map val (count list)) :list (conj list val)}]))
(defn rs-remove [{:keys [map list]} val]
  (if (not (contains? map val)) [false {:map map :list list}]
    (let [idx (map val) last-val (peek list)
          new-list (-> (pop list) (assoc idx last-val))
          new-map (-> (dissoc map val) (assoc last-val idx))]
      [true {:map new-map :list (if (= idx (dec (count list))) (pop list) new-list)}])))
